File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Emit/GenericNestedTypeInstanceReference.cs
using Microsoft.Cci;
using Microsoft.CodeAnalysis.Emit;
using Microsoft.CodeAnalysis.VisualBasic.Symbols;

namespace Microsoft.CodeAnalysis.VisualBasic.Emit
{
	internal sealed class GenericNestedTypeInstanceReference : GenericTypeInstanceReference, INestedTypeReference
	{
		public override IGenericTypeInstanceReference AsGenericTypeInstanceReference => this;

		public override INamespaceTypeReference AsNamespaceTypeReference => null;

		public override INestedTypeReference AsNestedTypeReference => this;

		public override ISpecializedNestedTypeReference AsSpecializedNestedTypeReference => null;

		public GenericNestedTypeInstanceReference(NamedTypeSymbol underlyingNamedType)
			: base(underlyingNamedType)
		{
		}

		private ITypeReference ITypeMemberReferenceGetContainingType(EmitContext context)
		{
			return ((PEModuleBuilder)context.Module).Translate(m_UnderlyingNamedType.ContainingType, (VisualBasicSyntaxNode)context.SyntaxNode, context.Diagnostics);
		}

		ITypeReference ITypeMemberReference.GetContainingType(EmitContext context)
		{
			//ILSpy generated this explicit interface implementation from .override directive in ITypeMemberReferenceGetContainingType
			return this.ITypeMemberReferenceGetContainingType(context);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Emit/GenericNestedTypeInstanceReference.vb
Imports Microsoft.Cci
Imports Microsoft.CodeAnalysis.Emit
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
	Friend NotInheritable Class GenericNestedTypeInstanceReference
		Inherits GenericTypeInstanceReference
		Implements INestedTypeReference
		Public Overrides ReadOnly Property AsGenericTypeInstanceReference As IGenericTypeInstanceReference
			Get
				Return Me
			End Get
		End Property

		Public Overrides ReadOnly Property AsNamespaceTypeReference As INamespaceTypeReference
			Get
				Return Nothing
			End Get
		End Property

		Public Overrides ReadOnly Property AsNestedTypeReference As INestedTypeReference
			Get
				Return Me
			End Get
		End Property

		Public Overrides ReadOnly Property AsSpecializedNestedTypeReference As ISpecializedNestedTypeReference
			Get
				Return Nothing
			End Get
		End Property

		Public Sub New(ByVal underlyingNamedType As NamedTypeSymbol)
			MyBase.New(underlyingNamedType)
		End Sub

		Private Function ITypeMemberReferenceGetContainingType(ByVal context As EmitContext) As ITypeReference
			Return DirectCast(context.[Module], PEModuleBuilder).Translate(Me.m_UnderlyingNamedType.ContainingType, DirectCast(context.SyntaxNode, VisualBasicSyntaxNode), context.Diagnostics, False, False)
		End Function
	End Class
End Namespace